lib/widgets/rectangular_inset_shadow_box.dart

```diff
@@ -0,0 +1,51 @@
+import 'package:flutter/widgets.dart';
+
+/// A widget that overlays rectangular inset shadows on a child.
+///
+/// The use case of this is casting shadow on scrollable UI elements.
+/// For example, when there is a list of items, the shadow could be
+/// a visual indicator for over scrolled areas.
+///
+/// See also:
+///  * https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=3860-11890&node-type=frame&t=oOVTdwGZgtvKv9i8-0
+class RectangularInsetShadowBox extends StatelessWidget {
+  const RectangularInsetShadowBox({
+    super.key,
+    this.top = 0,
+    this.bottom = 0,
+    required this.color,
+    required this.child,
+  });
+
+  /// The distance that the shadows from the child's top edge grows downwards.
+  ///
+  /// This does not pad the child widget.
+  final double top;
+
+  /// The distance that the shadows from the child's bottom edge grows upwards.
+  ///
+  /// This does not pad the child widget.
+  final double bottom;
+
+  /// The shadow color to fade into transparency from the top and bottom borders.
+  final Color color;
+
+  final Widget child;
+
+  BoxDecoration _shadowFrom(AlignmentGeometry begin) {
+    return BoxDecoration(gradient: LinearGradient(
+      begin: begin, end: -begin,
+      colors: [color, color.withValues(alpha: 0)]));
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(children: [
+      child,
+      Positioned(top: 0, left: 0, right: 0,
+        child: Container(height: top, decoration: _shadowFrom(Alignment.topCenter))),
```

Comment:
nit: Positioned can control the height, right?
```suggestion
      Positioned(top: 0, height: top, left: 0, right: 0,
        child: Container(decoration: _shadowFrom(Alignment.topCenter))),
```

Then also `Container` can I think be reduced to `DecoratedBox`.